chain/chunks/src/lib.rs

```diff
@@ -1266,11 +1276,17 @@ impl ShardsManager {
             // If we do care about the shard, we will remove the request once the full chunk is
             //    assembled.
             if !cares_about_shard {
+                let demurs_remain = self.seals_mgr.has_active_demur(&chunk_hash);
                 self.encoded_chunks.remove_from_cache_if_outside_horizon(&chunk_hash);
-                self.requested_partial_encoded_chunks.remove(&chunk_hash);
-                return Ok(ProcessPartialEncodedChunkResult::HaveAllPartsAndReceipts(
+                // We can only remove this from requested chunks if there are not parts
+                // we are waiting for to fulfil demurs.
+                if !demurs_remain {
```

Comment:
If they have remained, when would we remove the request?
It seems that the next time we receive the seals, we will not get to here, we will exit above when we check if all the parts and all the receipts are present (and we return `Known`), and we do not appear to remove the request there

Reply:
The request is removed on the very next line. This is why I added this `if` statement; we should not remove the request until enough seals have been fulfilled. It used to be the case that we did not enter this `have_all_parts && have_all_receipts` block while there were still active demurs, but since we are not blocking adding the chunk anymore now we do enter this block. However, we still cannot remove the chunk from the set of requests until after we have the seals too.

Unrelated note: please also comment on https://github.com/near/nearcore/pull/3657#issuecomment-738983558 when you have a chance.

Reply:
My question is specifically about the situation in which some demurs are still left here. We only remove the request on the next line if no demurs remain, if some remain, then we keep the request.

The next time the parts arrive, we will exit early on line 1159. There's new logic there that will process the seals, but not the logic that would remove the request, so it seems like the request will remain forever?

Reply:
Ah, ok. Sorry I misunderstood your question. Yes you are right that I was not properly removing the request if the seal was fulfilled after the other parts were known.

I have changed this in https://github.com/near/nearcore/pull/3657/commits/ca7556b5f4e0bfc77f8cee6e84cf6b540e0f55f3
It is a bit larger of a commit than strictly necessary to fix this issue, but I also refactored `track_seals` so that it does not iterate through all active seals on every call and I beefed up the unit testing to capture the requests getting removed.

Please let me know if you notice anything else wrong in the updated logic.